From 4e3fad33442dfd908561960921aad335fad6b5a8 Mon Sep 17 00:00:00 2001
From: Ray Strode <rstrode@redhat.com>
Date: Thu, 21 Jul 2011 16:09:13 -0400
Subject: [PATCH 4/4] daemon: call reload_users after timeout on passwd/shadow
 changes

Often the two files are changed in concert, so it makes sense to
wait a bit before doing a heavy reload operation.

Suggested by Vincent.

https://bugs.freedesktop.org/show_bug.cgi?id=39413
---
 src/daemon.c |   15 ++++++++++++++-
 1 files changed, 14 insertions(+), 1 deletions(-)

diff --git a/src/daemon.c b/src/daemon.c
index 882fb2f..79915a7 100644
--- a/src/daemon.c
+++ b/src/daemon.c
@@ -492,6 +492,19 @@ reload_autologin_timeout (Daemon *daemon)
 }
 
 static void
+queue_reload_users_soon (Daemon *daemon)
+{
+        if (daemon->priv->reload_id > 0) {
+                return;
+        }
+
+        /* we wait half a second or so in case /etc/passwd and
+         * /etc/shadow are changed at the same time, or repeatedly.
+         */
+        daemon->priv->reload_id = g_timeout_add (500, (GSourceFunc)reload_users_timeout, daemon);
+}
+
+static void
 queue_reload_users (Daemon *daemon)
 {
         if (daemon->priv->reload_id > 0) {
@@ -523,7 +536,7 @@ on_passwd_monitor_changed (GFileMonitor      *monitor,
                 return;
         }
 
-        reload_users (daemon);
+        queue_reload_users_soon (daemon);
 }
 
 static void
-- 
1.7.6

